service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ConsentMapper.java

```diff
@@ -0,0 +1,190 @@
+        POCDMT000002UK01StructuredBody structuredBody = getStructuredBody(clinicalDocument);
+        if (structuredBody != null) {
+            extractDataPeriodFromDoc(consent, structuredBody);
+            extractConsentSource(consent, structuredBody);
+            extractTextBody(consent, structuredBody);
+        }
+
+        return consent;
+    }
+
+    private void extractAuthCodesFromDoc(Consent consent, POCDMT000002UK01ClinicalDocument1 clinicalDocument) {
+        if (clinicalDocument.sizeOfAuthorizationArray() > 0) {
+            for (POCDMT000002UK01Authorization auth : clinicalDocument.getAuthorizationArray()) {
+                POCDMT000002UK01Consent authConsent = auth.getConsent();
+                CE ce = authConsent.getCode();
+                consent.addAction(getCodingFromCE(ce));
+            }
+        }
+    }
+
+    private void extractDataPeriodFromDoc(Consent consent, POCDMT000002UK01StructuredBody structuredBody) {
+        List<POCDMT000002UK01Entry> permissionEntries = getEntriesOfType(structuredBody, PERMISSION_TO_VIEW);
+        if (permissionEntries.isEmpty()) {
+            return;
+        }
+
+        for (POCDMT000002UK01Entry permissionEntry : permissionEntries) {
+            Period dataPeriod = getDataPeriod(permissionEntry);
+            if (dataPeriod == null) continue;
+            consent.setDataPeriod(dataPeriod);
+        }
+    }
+
+    private Period getDataPeriod(POCDMT000002UK01Entry permissionEntry) {
+        POCDMT000002UK01Observation observation = permissionEntry.getObservation();
+        if (observation == null || !observation.isSetEffectiveTime()) {
+            return null;
+        }
+
+        Period dataPeriod = new Period();
+        IVLTS effectiveTime = observation.getEffectiveTime();
+        if (effectiveTime.isSetLow()) {
+            dataPeriod.setStart(DateUtil.parse(effectiveTime.getLow().getValue()));
+        }
+        if (effectiveTime.isSetHigh()) {
+            dataPeriod.setEnd(DateUtil.parse(effectiveTime.getHigh().getValue()));
+        }
+        return dataPeriod;
+    }
+
+    private void extractTextBody(Consent consent, POCDMT000002UK01StructuredBody structuredBody) {
+        List<POCDMT000002UK01Section> sections = getSectionsOfType(structuredBody, ITK_SNOMED, SYSTEM_CODE);
+        for (POCDMT000002UK01Section section : sections) {
+            Narrative narrative = new Narrative();
+            narrative.setStatus(Narrative.NarrativeStatus.GENERATED);
+            narrative.setDivAsString(section.getText().xmlText());
+            consent.setText(narrative);
+        }
+    }
+
+    private void extractConsentSource(Consent consent, POCDMT000002UK01StructuredBody structuredBody) {
+        List<POCDMT000002UK01Section> sections = getSectionsOfType(structuredBody, ITK_SNOMED, SYSTEM_CODE);
+        for (POCDMT000002UK01Section section : sections) {
+            consent.setSource(new Identifier().setValue(section.getId().getRoot()));
```

Comment:
section.isSetId() check -> could cause NullPointerException

Reply:
added a check to prevent null pointer accordingly.

Reply:
changed to functional programming